Mathlib/Algebra/QuadraticAlgebra/Basic.lean

```diff
@@ -0,0 +1,389 @@
+  ext <;> simp
+
+end omega'
+
+section star
+
+variable [CommRing R]
+
+/-- Conjugation in `QuadraticAlgebra R a b`.
+The conjugate of `x + y ω` is `x + y ω' = (x - a * y) - y ω`. -/
+instance : Star (QuadraticAlgebra R a b) where
+  star z := ⟨z.re + b * z.im, -z.im⟩
+
+@[simp]
+theorem star_mk (x y : R) :
+    star (⟨x, y⟩ : QuadraticAlgebra R a b) = ⟨x + b * y, -y⟩ :=
+  rfl
+
+theorem star_omega : star (ω : QuadraticAlgebra R a b) = (ω') := by
+  simp [star, ω, ω']
+
+theorem star_omega' : star (ω' : QuadraticAlgebra R a b) = (ω) := by
+  simp [star, ω, ω']
+
+@[simp]
+theorem star_re (z : QuadraticAlgebra R a b) :
+    (star z).re = z.re + b * z.im :=
+  rfl
+
+@[simp]
+theorem star_im (z : QuadraticAlgebra R a b) :
+    (star z).im = -z.im :=
+  rfl
+
+theorem mul_star {x y : R} :
+    (⟨x, y⟩ * star ⟨x, y⟩ : QuadraticAlgebra R a b) =
+      x * x + b * x * y - a * y * y := by
+  ext <;> simp <;> ring
+
+instance : StarRing (QuadraticAlgebra R a b) where
+  star_involutive _ := by
+    refine QuadraticAlgebra.ext (by simp) (neg_neg _)
+  star_mul a b := by ext <;> simp <;> ring
+  star_add _ _ := QuadraticAlgebra.ext (by simp [star_re]; ring) (neg_add _ _)
+
+/-
+@[simp]
+theorem nsmul_val (n : ℕ) (x y : ℤ) :
+    (n : QuadraticAlgebra R a b) * ⟨x, y⟩ = ⟨n * x, n * y⟩ := by
+  ext <;> simp
+
+@[simp]
+theorem smul_val (n x y : ℤ) :
+    (n : QuadraticAlgebra R a b) * ⟨x, y⟩ = ⟨n * x, n * y⟩ := by
+  ext <;> simp
+
+@[simp]
+theorem smul_re (r : R) (z : QuadraticAlgebra R a b) :
+    (r • z).re = r * z.re := by
+  simp
+
+@[simp]
+theorem smul_im (r : R) (z : QuadraticAlgebra R a b) :
+    (r • z).im = r * z.im := by
+  simp
+-/
+
+end star
+
+section norm
+
+variable [CommRing R]
+
+/-- The norm of an element of `QuadraticAlgebra R a b`. -/
+def norm (z : QuadraticAlgebra R a b) : R :=
+  z.re * z.re + b * z.re * z.im - a * z.im * z.im
+
+theorem norm_def (z : QuadraticAlgebra R a b) :
+    z.norm = z.re * z.re + b * z.re * z.im - a * z.im * z.im :=
+  rfl
+
+@[simp]
+theorem norm_zero : norm (0 : QuadraticAlgebra R a b) = 0 := by simp [norm]
+
+@[simp]
+theorem norm_one : norm (1 : QuadraticAlgebra R a b) = 1 := by simp [norm]
+
+@[simp]
+theorem norm_coe (r : R) : norm (r : QuadraticAlgebra R a b) = r * r := by simp [norm_def]
+
+@[simp]
+theorem norm_natCast (n : ℕ) : norm (n : QuadraticAlgebra R a b) = n * n :=
+  by simp [norm_def]
+
+@[simp]
+theorem norm_intCast (n : ℤ) : norm (n : QuadraticAlgebra R a b) = n * n :=
+  by simp [norm_def]
+
+@[simp]
+theorem norm_mul (z w : QuadraticAlgebra R a b) :
+    norm (z * w) = norm z * norm w := by
+  simp [norm_def]; ring
+
+/-- `norm` as a `MonoidHom`. -/
+def normMonoidHom : QuadraticAlgebra R a b →* R where
+  toFun := norm
+  map_mul' := norm_mul
+  map_one' := norm_one
+
+theorem normMonoidHom_apply (r : QuadraticAlgebra R a b) :
+    normMonoidHom r = norm r := rfl
+
+theorem coe_norm_eq_mul_conj (z : QuadraticAlgebra R a b) :
```

Comment:
I would suggest sticking to the `star` naming instead of `conj`, unless there is a plan to introduce `conj` in some scope as we do for `Complex`. But I wouldn't recommend that.
```suggestion
theorem coe_norm_eq_mul_star (z : QuadraticAlgebra R a b) :
```